text/0000-generalized-type-ascription.md

```diff
@@ -0,0 +1,1740 @@
+    println!("{:?}", (0..10).map(|x| x % 3 == 0).collect());
+
+    let _ = Box::new("1234".parse().unwrap());
+}
+```
+
+but get two errors of form:
+
+```rust
+error[E0283]: type annotations required: ...
+
+error[E0282]: type annotations needed
+```
+
+you can resolve the errors by writing:
+
+```rust
+#![feature(type_ascription)]
+
+fn main() {
+    println!("{:?}", (0..10).map(|x| x % 3 == 0).collect() : Vec<_>);
+
+    let _ = ("1234".parse().unwrap() : usize).into() : Box<_>;
+}
+```
+
+-----------------
+
+*Aside:* You can also resolve the above errors by using turbofish and `Box::new`:
+
+```rust
+fn main() {
+    println!("{:?}", (0..10).map(|x| x % 3 == 0).collect::<Vec<_>>());
+
+    let x = Box::new("1234".parse::<usize>().unwrap());
+}
+```
+
+### In macros
+
+Note that the fact that `expr : Type` is a valid expression extends to macros
+as well and this is implemented in a nightly compiler right now.
+For example, we can make and invoke macro that ascribes an expression with a
+type `Vec<$t>` with the following valid snippet:
+
+```rust
+#![feature(type_ascription)]
+
+macro_rules! ascribe {
+    ($e: expr, $t: ty) => {
+        $e : Vec<$t>
+    }
+}
+
+fn main() {
+    let _ = ascribe!(vec![1, 2, 3], u8);
+}
+```
+
+### Precedence of the operator
+
+[prec_unresolved]: https://github.com/rust-lang/rfcs/blob/master/text/0803-type-ascription.md#unresolved-questions
+
+[prec_ref]: https://doc.rust-lang.org/beta/reference/expressions.html
+
+[RFC 803] proposed and implemented that `:` as an operator in expression
+contexts should have the same precedence as the `as` operator
+*([see the reference][prec_ref])*. However, the RFC also
+[left this question unresolved][prec_unresolved] and asked:
+
+> Is the suggested precedence correct?
+
+We argue in this RFC that the current implementation is sub-optimal and thus
+propose that the precedence should be slightly changed.
+
+To see why, consider the example above where we wrote:
+
+```rust
+let _ = ("1234".parse().unwrap() : usize).into() : Box<_>;
+```
+
+Notice in particular here that we've had to enclose the inner ascription
+in parenthesis. Consider that you are writing this snippet and reach the
+`.into()`. Once you do that, you'll need to select everything on the line
+until before the `=` token. This can slow down your writing flow.
+Furthermore, as we chain more and more methods, the build-up of parenthesis
+can increase and thus make writing and reading further impaired. An example:
+
+```rust
+let x = (((0..10)
+    .map(some_computation)
+    .collect() : Result<Vec<_>, _>)
+    .unwrap()
+    .map(other_computation) : Vec<usize>)
+    .into() : Rc<[_]>;
+```
+
+We suggest instead that you should be able to write:
+
+```rust
+let x = (0..10)
+    .map(some_computation)
+    .collect() : Result<Vec<_>, _>
+    .unwrap()
+    .map(other_computation) : Vec<usize>
+    .into() : Rc<[_]>;
+```
+
+To that end, `foo : bar.quux()` and `foo : bar.quux` should unambiguously be
```

Comment:
Note that this is unintuitive if Rust ever has type-level functions, as there one would naturally expect `bar.quux()` to be associated the most tightly.

Reply:
Thanks; added a section in the rationale discussing this.